Runtime/Rust/src/rpc/router/mod.rs

```diff
@@ -0,0 +1,139 @@
+
+use crate::rpc::transport::TransportProtocol;
+use crate::rpc::{Datagram, DynFuture};
+
+pub use self::calls::{CallDetails, RequestHandle, TypedRequestHandle};
+use self::context::{SpawnTask, UnknownResponseHandler};
+
+mod call_table;
+pub mod calls;
+mod context;
+
+/// The local end of the pipe handles messages. Implementations are automatically generated from
+/// bebop service definitions.
+///
+/// You should not implement this by hand but rather the *Handlers traits which are generated.
+pub trait ServiceHandlers: Send + Sync {
+    /// The name of this service.
+    ///
+    /// This is a const for a given implementation, however for dynamic type support it has to be
+    /// provided via a function.
+    fn _name(&self) -> &'static str;
+
+    /// Use opcode to determine which function to call, whether the signature matches,
+    /// how to read the buffer, and then convert the returned values and send them as a
+    /// response
+    ///
+    /// This should only be called by the `Router` and is not for external use.
+    ///
+    /// This returns a future instead of being async because it must decode datagram before starting
+    /// the async section.
+    #[doc(hidden)]
+    fn _recv_call<'f>(&self, datagram: &Datagram, handle: RequestHandle) -> DynFuture<'f>;
+}
+
+// impl<T: Deref<Target = S> + Send + Sync, S: ServiceHandlers> ServiceHandlers for T {
+//     fn _name(&self) -> &'static str {
+//         self.deref()._name()
+//     }
+//
+//     fn _recv_call<'f>(&self, datagram: &Datagram, transport: RequestHandle) -> DynFuture<'f> {
+//         self.deref()._recv_call(datagram, transport)
+//     }
+// }
+
+assert_obj_safe!(ServiceHandlers);
+
+/// Wrappers around the process of calling remote functions. Implementations are generated from
+/// bebop service definitions.
+///
+/// You should not implement this by hand.
+pub trait ServiceRequests: Send + Sync {
+    const NAME: &'static str;
+
+    fn new(ctx: Weak<RouterContext>) -> Self;
+}
+
+impl<D, S> ServiceRequests for D
+where
+    D: Deref<Target = S> + Send + Sync,
+    S: ServiceRequests + Into<D>,
+{
+    const NAME: &'static str = S::NAME;
+
+    fn new(ctx: Weak<RouterContext>) -> Self {
+        S::new(ctx).into()
+    }
+}
+
+/// This is the main structure which represents information about both ends of the connection and
+/// maintains the needed state to make and receive calls. This is the only struct of which an
+/// instance should need to be maintained by the user.
+pub struct Router<Remote> {
+    /// Remote service converts requests from us, so this also provides the callable RPC functions.
+    _remote_service: Remote,
+
+    /// Router state that can be referenced by multiple places. This should only be used by
+    /// generated code.
+    ///
+    /// **Warning:** always store weak references or the router will have issues shutting down due
+    /// to cyclical dependencies.
+    #[doc(hidden)]
+    pub _context: Arc<RouterContext>,
```

Comment:
Can we not change the type to `Weak<RouterContext>` ?

Reply:
No, this is the only owner.